File: src/api/gameCategories.ts
import { invoke } from '@tauri-apps/api/core';
import { z } from 'zod';
import {
  GameCategorySchema,
  UpsertGameCategoryRequestSchema,
  type GameCategory,
  type UpsertGameCategoryRequest,
} from '../schemas';

/**
 * 全ゲームカテゴリを取得
 */
export async function getGameCategories(): Promise<GameCategory[]> {
  const result = await invoke<unknown>('get_game_categories');
  return z.array(GameCategorySchema).parse(result);
}

/**
 * IDでゲームカテゴリを取得
 */
export async function getGameCategory(gameId: string): Promise<GameCategory | null> {
  const result = await invoke<unknown>('get_game_category', { gameId });
  return result === null ? null : GameCategorySchema.parse(result);
}

/**
 * ゲームカテゴリを挿入または更新
 */
export async function upsertGameCategory(request: UpsertGameCategoryRequest): Promise<void> {
  const validatedRequest = UpsertGameCategoryRequestSchema.parse(request);
  await invoke('upsert_game_category', { request: validatedRequest });
}

/**
 * ゲームカテゴリを削除
 */
export async function deleteGameCategory(gameId: string): Promise<void> {
  await invoke('delete_game_category', { gameId });
}

/**
 * ゲームカテゴリを検索
 */
export async function searchGameCategories(query: string): Promise<GameCategory[]> {
  const result = await invoke<unknown>('search_game_categories', { query });
  return z.array(GameCategorySchema).parse(result);
}

File: src-tauri/src/commands/game_categories.rs
use crate::database::{
    models::GameCategory, repositories::GameCategoryRepository, DatabaseManager,
};
use serde::{Deserialize, Serialize};
use tauri::State;

/// ゲームカテゴリの挿入/更新リクエスト
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpsertGameCategoryRequest {
    pub game_id: String,
    pub game_name: String,
    pub box_art_url: Option<String>,
}

/// 全ゲームカテゴリを取得
#[tauri::command]
pub async fn get_game_categories(
    db_manager: State<'_, DatabaseManager>,
) -> Result<Vec<GameCategory>, String> {
    let conn = db_manager
        .get_connection()
        .await
        .map_err(|e| format!("Failed to get database connection: {}", e))?;

    GameCategoryRepository::get_all_categories(&conn)
        .map_err(|e| format!("Failed to get game categories: {}", e))
}

/// IDでゲームカテゴリを取得
#[tauri::command]
pub async fn get_game_category(
    db_manager: State<'_, DatabaseManager>,
    game_id: String,
) -> Result<Option<GameCategory>, String> {
    let conn = db_manager
        .get_connection()
        .await
        .map_err(|e| format!("Failed to get database connection: {}", e))?;

    GameCategoryRepository::get_category_by_id(&conn, &game_id)
        .map_err(|e| format!("Failed to get game category: {}", e))
}

/// ゲームカテゴリを挿入または更新
#[tauri::command]
pub async fn upsert_game_category(
    db_manager: State<'_, DatabaseManager>,
    request: UpsertGameCategoryRequest,
) -> Result<(), String> {
    let conn = db_manager
        .get_connection()
        .await
        .map_err(|e| format!("Failed to get database connection: {}", e))?;

    GameCategoryRepository::upsert_category(
        &conn,
        &request.game_id,
        &request.game_name,
        request.box_art_url.as_deref(),
    )
    .map_err(|e| format!("Failed to upsert game category: {}", e))
}

/// ゲームカテゴリを削除
#[tauri::command]
pub async fn delete_game_category(
    db_manager: State<'_, DatabaseManager>,
    game_id: String,
) -> Result<(), String> {
    let conn = db_manager
        .get_connection()
        .await
        .map_err(|e| format!("Failed to get database connection: {}", e))?;

    GameCategoryRepository::delete_category(&conn, &game_id)
        .map_err(|e| format!("Failed to delete game category: {}", e))
}

/// ゲームカテゴリを検索
#[tauri::command]
pub async fn search_game_categories(
    db_manager: State<'_, DatabaseManager>,
    query: String,
) -> Result<Vec<GameCategory>, String> {
    let conn = db_manager
        .get_connection()
        .await
        .map_err(|e| format!("Failed to get database connection: {}", e))?;

    GameCategoryRepository::search_categories(&conn, &query)
        .map_err(|e| format!("Failed to search game categories: {}", e))
}
